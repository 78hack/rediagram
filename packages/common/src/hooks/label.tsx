import React, { ReactElement, isValidElement, ReactNode, useMemo, Fragment } from 'react';
import { DOT } from '@ts-graphviz/react';

function transformToHtmlLike(value: string): string | ReactElement {
  if (value.includes('\\n')) {
    return (
      <>
        {value.split('\\n').map((v) => (
          <Fragment key={v}>
            {v}
            <DOT.BR />
          </Fragment>
        ))}
      </>
    );
  }
  return value;
}

export function useLabelText(
  label?: string | ReactNode,
  { defaultValue, htmlLike = false }: { defaultValue?: string; htmlLike?: boolean } = {},
): string | ReactElement | undefined {
  return useMemo(() => {
    if (isValidElement(label)) {
      return label;
    }
    if (typeof label === 'string') {
      return htmlLike ? transformToHtmlLike(label) : label;
    }
    return htmlLike && defaultValue ? transformToHtmlLike(defaultValue) : defaultValue;
  }, [label, htmlLike, defaultValue]);
}
