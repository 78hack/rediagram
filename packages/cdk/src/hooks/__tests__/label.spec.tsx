import React from 'react';
import { DOT } from '@ts-graphviz/react';
import { renderHook } from '@testing-library/react-hooks';

import { useLabelText } from '../label';

describe('label', () => {
  describe('useLabelText', () => {
    it('should return as it is when input ReactElement', () => {
      const { result } = renderHook(() => useLabelText(<DOT.I>test</DOT.I>));
      expect(result.current).toMatchSnapshot();
    });
  });
});
