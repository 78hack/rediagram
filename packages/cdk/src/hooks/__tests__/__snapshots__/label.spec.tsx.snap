// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`label useLabelText should return as it is when input ReactElement 1`] = `
<dot-i>
  test
</dot-i>
`;
